src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,279 @@
+    return -EINVAL;
+  }
+  bucket_name = bucket_name_marker.substr(0, pos);
+  marker = bucket_name_marker.substr(pos + 1);
+
+  return 0;
+}
+
+// migrate v1 notification metadata for a single bucket
+int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::RadosStore* driver, const rgw_raw_obj& obj)
+{
+  // parse bucket name and marker of out "pubsub.{tenant}.bucket.{name}/{marker}"
+  auto* rados = driver->getRados()->get_rados_handle();
+  std::string tenant;
+  std::string bucket_name;
+  std::string marker;
+  int r = deconstruct_topics_oid(obj.oid, tenant, bucket_name, marker, dpp);
+  if (r < 0) {
+    std::string s = "failed to read tenant, bucket name and marker from " + obj.to_str()
+                  + " with: " + cpp_strerror(r)
+                  + " expected format pubsub.{tenant}.bucket.{name}/{marker}!";
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to read v1 bucket notifications from " + obj.to_str()
+                  + " with: " + cpp_strerror(r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // break if RGW_ATTR_BUCKET_NOTIFICATION xattr already exists
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  r = -ECANCELED;
+  for (auto i = 0u; i < 15u && r == -ECANCELED; ++i) {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r == -ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      std::string s = "failed to load the bucket from " + obj.to_str()
+                    + " with: " + cpp_strerror(r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+
+    if (bucket->get_marker() != marker) {
+      break;
+    }
+
+    bufferlist bl;
+    bucket_topics.encode(bl);
+    // same attrs ready to merge
+    rgw::sal::Attrs attrs = bucket->get_attrs();
+
+    if(bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+      // change only bucket notification attr
+      attrs[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+      r = bucket->merge_and_store_attrs(dpp, attrs, y);
+    } else {
+      bucket->get_attrs()[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+      r = bucket->put_info(dpp, false, real_time(), y);
+    }
```

Comment:
So the question is do we want to migrate the v1 notifications, if we have already have v2 notification on the bucket?
we can disable and enable the feature anytime, so do we want to migrate only once on a bucket or whenever this happens
this is the idea of this code, although I have simpler idea to do it but havn't pushed it yet because it I am still debugging

Reply:
it's okay for users to disable v2 if they need to work around migration issues, but we don't want to support toggling v2 on and off randomly

we just want the migration to be as simple and error-free as possible. if migration has already run and generated v2 metadata, running migration again should not overwrite any existing v2 metadata